lib/lambda_ethereum_consensus/state_transition/cache.ex

```diff
@@ -0,0 +1,91 @@
+defmodule LambdaEthereumConsensus.StateTransition.Cache do
+  @moduledoc """
+  Caches expensive function calls.
+  """
+
+  alias LambdaEthereumConsensus.StateTransition.{Accessors, Misc}
+
+  @spec init_cache_tables() :: :ok
+  def init_cache_tables do
+    [:total_active_balance, :beacon_proposer_index, :beacon_committee, :active_validator_count]
+    |> Enum.each(fn table ->
+      :ets.new(table, [:set, :public, :named_table])
+    end)
+  end
+
+  defp lookup(table, key, compute_fun) do
+    case :ets.lookup(table, key) do
+      [{^key, balance}] ->
+        balance
+
+      [] ->
+        balance = compute_fun.()
+        :ets.insert_new(table, {key, balance})
+        balance
+    end
+  end
+
+  @spec cache_total_active_balance(SszTypes.BeaconState.t(), (-> SszTypes.gwei())) ::
+          SszTypes.gwei()
+  def cache_total_active_balance(state, compute_fun) do
+    epoch = Accessors.get_current_epoch(state)
+    root = get_epoch_root(state, epoch)
+
+    lookup(:total_active_balance, {epoch, root}, compute_fun)
+  rescue
+    _ -> compute_fun.()
+  end
+
+  @spec cache_beacon_proposer_index(SszTypes.BeaconState.t(), (-> SszTypes.validator_index())) ::
+          SszTypes.validator_index()
+  def cache_beacon_proposer_index(%SszTypes.BeaconState{slot: slot} = state, compute_fun) do
```

Comment:
couldn't you just pass the lookup keys to a generic function? 

Like:

```
Cache.cache_fun(:beacon_proposer_index, {slot, root}, fn ->
      epoch = get_current_epoch(state)

      indices = get_active_validator_indices(state, epoch)

      state
      |> get_seed(epoch, Constants.domain_beacon_proposer())
      |> then(&SszEx.hash(&1 <> Misc.uint64_to_bytes(state.slot)))
      |> then(&Misc.compute_proposer_index(state, indices, &1))
    end)
)
```

Reply:
Yeah, it's cleaner like that.